kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/NumberedPolynomial.kt

```diff
@@ -0,0 +1,427 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+@file:Suppress("NOTHING_TO_INLINE", "KotlinRedundantDiagnosticSuppress")
+
+package space.kscience.kmath.functions
+
+import space.kscience.kmath.operations.Ring
+import space.kscience.kmath.structures.Buffer
+import kotlin.jvm.JvmName
+import kotlin.math.max
+
+
+/**
+ * Represents multivariate polynomial that stores its coefficients in a [Map] and terms' signatures in a [List].
+ *
+ * @param C the type of constants.
+ */
+public data class NumberedPolynomial<C>
+@PublishedApi
+internal constructor(
+    /**
```

Comment:
Actually it depends on what level of safety you want to provide. There is a problem in representation of polynomial that type system does not cover: signature of monomial can have extra zeros at the end, which allows (from type system's POV) to declare two different monomials with the same signature. For example, `5x_0 = 2x_0 + 3x_0` can be represented as
```kotlin
mapOf(
    listOf(1u) to 5,
)
```
as well as
```kotlin
mapOf(
    listOf(1u) to 2,
    listOf(1u, 0u) to 3,
)
```
So I see only such solutions of the problem:
1. Ignore it. ("It's not a bug, it's a feature.")
2. Prohibit constructing incorrect numbered polynomials by controlling the way they are constructed.
3. Allow constructing incorrect numbered polynomials and fix them in arithmetic operations.

I assumed that way # 2 is the best. So I closed the constructor and made:
1. fabric that checks and fixes received coefficients map,
2. and functions (see `NumberedPolynomialAsIs` and `NumberedPolynomialWithoutCheck`) that make the constructor implicitly public but warn that there could be problems in case of their misuse.